[2단계 - 로또 구현] 김김(김기문) 미션 제출합니다.

tributetothemoon opened the pull request:
안녕하세요 데이브! 또 뵙겠습니다 :)

저번 [리뷰](https://github.com/woowacourse/java-lotto/pull/261)에서 남겨주신 코멘트를 반영하고(반영 커밋은 링크에 코멘트로 달아두었습니다!), 2단계 미션을 진행하였습니다.

원래 1차 리뷰 신청 기한이 생기면서 급히 구현하다보니 리팩토링에 대한 필요성을 느꼈었기도 하였고, 2단계 미션도 진행하면서 새로운 기능들이 추가되어야 했기에, 대대적인 리팩토링을 진행하였습니다.

그래서 굉장히 많이 바뀐만큼 어떻게 변경되었느냐를 설명하는 것보다 도메인 클래스들의 역할을 적는게 나을 것 같아 정리해보았습니다!

### `LottoNumber`
- 1자리의 로또 번호를 뜻하는 클래스입니다.

### `LottoTicket`
- 1장의 로또(6개의 로또 번호)를 뜻하는 클래스입니다.

### `WinningNumbers`
- 로또의 당첨번호를 뜻합니다.
- 로또의 당첨 등수 계산에는 6자리의 번호(LottoTicket)와 1자리의 보너스볼(LottoNumber)이 쓰이기 때문에, 인스턴스로 LottoTicket과 LottoNumber를 가집니다.
- calculateRank(LottoTicket)메소드는 외부의 로또 티켓을 받아 자신의 6자리 정답 번호와 보너스 볼 등을 감안하여 등수를 계산하여 리턴합니다. 즉, 등수 계산은 우승 번호에 대한 정보를 가진 WinningNumbers객체가 외부의 로또 객체를 인자로 받아 행할 수 있습니다.

### `LottoNumberRepository`
- 자동으로 로또 번호를 만들어내는 데에 사용하기 위해 만들었습니다. 로또 무작위 자동 생성이 실행될 때 마다 매번 45개의 번호를 생성하고 이를 섞는 것은 다소 비효율적이라고 생각했기 때문입니다.

### `Rank`
- 로또의 등수들을 정의한 enum클래스입니다. 인스턴스 변수로 몇개를 맞춰야 하는지와, 상금은 얼마인지를 담습니다.

### `Money`
- 프로그램 내에서 사용되는 돈을 뜻합니다.

### `LottoPurchaseManager`
- 사용자가 입력한 구입 금액에 해당하는 Money 객체를 받고 생성됩니다.
- 자동/수동 로또 구매 메소드가 있으며, 해당 메소드가 불릴 때마다 로또의 가격만큼 Money를 차감하고 로또 티켓을 생성합니다.
- 현재 자금 보유량에 따라 로또를 구매할 수 있는지 역시 판별하여 boolean으로 리턴 할 수 있습니다.

### `LottoResult`
- 로또 당첨 통계(e.g., 1등 몇개, 2등 몇개 등등)를 담기 위해 만들어졌습니다.
- 정적 팩토리 메소드 of를 통해 우승 번호(WinningNumbers) 객체와 사용자가 구입한 로또 티켓들의 리스트(List<LottoTicket>)을 받고 생성됩니다.
- 정적 팩토리 과정에서 당첨 통계를 계산하여 Map으로 담습니다.
- getReward()메소드는 Map에 담긴 당첨 통계를 이용해 상금을 계산해 리턴해줍니다.

### `Profit`
- 수익률을 뜻합니다.
- 정적 팩토리 메소드의 인자로 초기 투입금과 상금을 받고, 이를 통해 계산하여 생성됩니다.

### `Repeater`
저번 미션과 코드 구조상으로 크게 달라진 부분은, util 패키지에 만든 `Repeater`라는 클래스로 인한 리팩토링인데요, 이 유틸리티 클래스를 만든 이유는, 기존에는 사용자 입력에서 제대로 된 값이 들어올 때까지 재입력을 강제하고 싶을 때 마다  try ~ catch문을 이용하여 재귀호출을 사용했는데, try~ catch문이 복잡하게 중첩되고, 또한 같은 코드가 중복되는 것이 보기 싫었기 때문입니다.
그래서 Repeater.repeatFunctionOnError(Supplier<T>)메소드는 함수를 받아 그 함수를 실행하고, 중간에 예외가 발생하면 에러 메세지를 출력한 뒤 다시 해당 함수를 반복하는 기능을 수행합니다.
결론적으로는 예외 발생 시 반복이라는 기능을 간단히 함수를 넘겨주는 것만으로 구현 할 수 있어 편리했지만, 이것 나름대로 코드가 더러워지는 현상이 발생한 것 같습니다... 차라리 try~catch문을 쓰는게 협업의 측면에선 나은지 나름대로 고민을 남기는 클래스였습니다.

----------

이상 김김이었습니다. 다시 한 번, 바쁘신 와중에 리뷰를 해주셔서 대단히 감사합니다 데이브!

pobiconan committed:
refactor: remove spark library, html

tributetothemoon committed:
style: 오타 일괄 수정

tributetothemoon committed:
style: 테스트명 구체화

tributetothemoon committed:
style: WinningNumber -> WinningNumbers 복수형으로 일관화

tributetothemoon committed:
feat: long 타입의 값을 받아 Money를 생성할 수 있는 생성자 추가

tributetothemoon committed:
refactor: 내부 로직 및 테스트 시 long값을 이용해 Money 객체를 생성토록 변경

tributetothemoon committed:
refactor: Rank 내부 어트리뷰트 private final로 속성 변경

tributetothemoon committed:
refactor: WinningNumbersTest Fixture를 이용해 리팩토링

tributetothemoon committed:
refactor: 정규표현식 Pattern 객체를 통해 성능 개선

- 기존의 String.matches()는 매번 컴파일 되어 객체를 만드는 데 이는 성능상 매우 비효율적이기에, Pattern객체를 상수화하여 로드타임에 생성하여 사용합니다.

tributetothemoon committed:
feat: Money 클래스에 뺄셈과 곱셈 연산 함수 추가

tributetothemoon committed:
feat: 자금을 관리하는 Budget 클래스 생성

tributetothemoon committed:
refactor: Controller에서 개행 메소드 삭제

tributetothemoon committed:
refactor: LottoTickets 삭제

- 별 다른 역할이 없는듯하여 삭제하였습니다.
- 기존의 LottoTickets가 필요한 기능은 모두 List<LottoTicket>를 통해 해결하도록 대체되었습니다.

tributetothemoon committed:
refactor: Budget을 적용해 로또 구매 로직 변경

tributetothemoon committed:
refactor: 당첨번호 검증 과정에서 적절한 구분자인지 검증하는 과정 삭제

- 구분자를 쉼표로 입력하지 않은 경우는 그 자체로 경우의 수가 매우 많기에 이를 인식하기 위한 검증 로직을 별도로 두기에는 다소 무리가 있습니다. 하지만 어떠한 경우든, 콤마로 제대로 구분하지 않은 입력은 도메인의 검증로직(넘겨진 리스트 요소의 개수, 숫자가 아닌 문자 포함 등에 대한 검증)으로 인해 예외가 발생하고 재입력을 요구하기 때문에 구분자 검증 로직은 별도로 두지 않아도 정상적인 프로그램의 실행이 가능합니다. 이에 삭제합니다.

tributetothemoon committed:
refactor: InputView 자체 Integer변환 로직 적용

- Integer가 필요한 경우 기존의 로직은 InputView에서 String을 넘겨받고 도메인에서 Integer로 변환하도록 구성되어 있었습니다. Integer로 변환하는 로직을 View에서 담당하도록 변경하였습니다.
- 동시에 Int로 형변환하는 도메인 로직은 전부 삭제되었고, 따라서 int를 받는 생성자들을 추가하였습니다.
- 형변환 및 공백 등 InputView에서 검증하는 과정에서 예외 발생시, 재입력을 요구하는 과정을 ScanRepeater를 이용해 리팩토링하였습니다.

tributetothemoon committed:
refactor: ScanRepeater -> Repeater로 변경. 컨트롤 로직에 활용

- 컨트롤 로직에서도 충분히 활용 가능한 메소드를 담고 있었기 때문에 이를 컨트롤러에도 적용했습니다.
- 동시에 View뿐만 아니라 Controller에도 적용하는 유틸리티의 성격이 강하므로 패키지를 변경하였습니다.

tributetothemoon committed:
refactor: OutputView import문 개선

tributetothemoon committed:
refactor: import문 최적화

tributetothemoon committed:
feat: View 메소드들에 개행 로직 추가

tributetothemoon committed:
feat: 수동 로또 구매 기능 구현

tributetothemoon committed:
refactor: Budget 삭제, LottoPurchase로 통합

- 기존에는 LottoPurchase의 메소드들이 Budget을 파라미터로 받고, 그 Budget에서 금액을 차감하는 방식이었는데, 이는 하나의 Side-Effect를 발생시키는 것과 같았습니다. 따라서 이를 수정할 필요가 있었습니다.
- 이에 static 메소드를 통한 유틸리티 클래스였던 LottoPurchase는 이제 객체로 생성되며, 생성자로 예산(Money)을 받습니다.
- 구매 메소드가 실행될 때 마다 예산을 차감하며 로또를 구매하는 역할을 하게 됩니다.
- Budget의 역할은 Purchase가 흡수하게 되면서 삭제되었습니다.

tributetothemoon committed:
style: LottoPurchase -> LottoPurchaseManager 네이밍 변경

tributetothemoon committed:
refactor: 문자열을 이용한 Money 생성자 삭제

- InputView에서 숫자 입력이 아닐 시 재입력을 요구할 수 있는 기능이 충분히 마련되어있기에, 도메인 내부에서는 String을 받지 않도록 합니다.

tributetothemoon committed:
refactor: Statistics -> LottoResult 변경

- 통계라는 단어만으로는 무엇에 대한 통계인지 알 수 없다고 판단하여 네이밍을 변경합니다.
- OutputView.printStatistics의 네이밍 역시 변경되었으며, 또한 파라미터 타입이 맵(Rank, Integer)으로 변경되었습니다. 이는, View단에는 필요한 데이터만을 간추려 DTO로 전송한다는 데에서 착안합니다.

tributetothemoon committed:
style: 코드 포맷팅

tributetothemoon committed:
refactor: LottoResult 정적 팩토리 메소드를 통해 생성되도록 변경

- 정적 팩토리 메소드를 통해 만들어지도록 변경하였습니다.
- 팩토리 메소드 과정에서 당첨 통계값을 구해 Map을 구성합니다.
- 기본 생성자에서 unmodifiable 콜렉션을 이용하여 할당토록하여 불변성을 보장하도록 하였습니다.

tributetothemoon committed:
refactor: 모든 도메인 객체가 정적 팩토리 메소드를 통해 생성되도록 변경

- 기존의 경우 어떤 것은 정적 팩토리 메소드로 생성되지만 어떠한 것은 생성자로 생성되기에 일관성이 없다고 판단했습니다. 이에 모두 정적 팩토리를 통해서만 생성되도록 변경합니다.

tributetothemoon committed:
style: LottoTicket.of의 파라미터 명 변경

- 섞인 로또 번호들인지 아닌지는 LottoTicket의 관심사가 아닙니다. 따라서 변경합니다.

dave915 commented on src/main/java/util/Repeater.java:
```diff
@@ -0,0 +1,18 @@
+public class Repeater {
+    private static OutputView outputView = OutputView.getInstance();
+
+    public static <T> T repeatFunctionOnError(Supplier<T> supplier) {
+        try {
+            return supplier.get();
+        } catch (RuntimeException e) {
+            outputView.printError(e);
+            return repeatFunctionOnError(supplier);
+        }
+    }
```
👍

tributetothemoon replied:
😆🙌

dave915 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,43 +1,72 @@
+
+    private List<LottoTicket> buyLottosManually(LottoPurchaseManager lottoPurchaseManager) {
+        int quantity = Repeater.repeatFunctionOnError(inputView::scanManualLottoQuantity);
+        if (!lottoPurchaseManager.canAfford(LottoTicket.PRICE, quantity)) {
```
LottoPurchaseManger내부에서도 LottoTicket.PRICE 가 쓰이고 있기 때문에 canAfford 호출시 추가로 안넘겨주어도 되지 않을까요?

tributetothemoon replied:
[수정](https://github.com/tributetothemoon/java-lotto/commit/3ef6b1c675cbf8877a8bb2bdbabdaabbfe3baf2a)하였습니다. LottoPurchaseManager 객체인 만큼 어차피 로또 몇개를 살 수 있는가만 판단하면 되는 문제였는데, 괜히 아이템의 단위 가격을 받고 있었네요. 좋은 피드백 감사합니다!

dave915 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,43 +1,72 @@
-        } catch (RuntimeException e) {
-            outputView.printError(e);
-            return createBudgetMoney();
+    private List<LottoTicket> buyLottosAutomatically(LottoPurchaseManager lottoPurchaseManager) {
```
`List<LottoTicket>`에 일급컬렉션을 적용해 로또 생성 로직과 `List<LottoTicket>`를 사용하는 로직들을 일급컬렉션 내부로 이동시켜보면 어떨까요?

tributetothemoon replied:
사실 step1에서는 만들었다가 삭제([커밋내역](https://github.com/tributetothemoon/java-lotto/commit/20bdd5b8f5dd7ed353fad8a339d33dc6d9ee13c9))하였는데요... 사실상 tickets를 만들고 나니 리스트에서 가지고 있는 메소드(size()나 티켓 리스트를 unmodifiableList로 리턴 하는 lottoTickets()라든가...)만을 추가로 구현하게 되길래 만드는 의미가 없는 것 같아 삭제하였습니다. 만약 일급컬렉션을 적용한다면, `List<LottiTickets>`에 대한 일급 컬렉션 `LottoTickets`는 무슨 역할을 할 수 있도록 구현해야할까요?

dave915 replied:
일급컬렉션을 적용하여 LottoTickets를 생성하는 책임, LottoTickets에게 winningNumber를 전달하여 LottoResult를 반환하도록 하는 책임등을 부여 할 수 있지 않을까요?

tributetothemoon replied:
[커밋링크](https://github.com/tributetothemoon/java-lotto/commit/98cf5352b30f88581020eef38ad0704cb541da1d)
우선은 List<LottoTicket>을 사용하는 모든 로직을 위 커밋과 같이 LottoTickets 클래스를 생성하여 Wrapping해보았습니다. 그 결과 코드의 가독성도 상승한 것 같고, 외부에서는 List 인터페이스를 알 필요 없이 프로그램의 로직을 구현할 수 있어 더 직관적이게 되었다는 생각이 들었습니다.
LottoTicket의 생성은 구입 자금과 결부되어있어 LottoPurchaseManager에 위임한 상태지만, 이를 통해 만들어지는 LottoTicket들을 LottoTickets에 관리하도록 위임함으로써 유저가 구매한 모든 티켓에 대한 관리가 용이해짐을 느낄 수 있었던 것 같습니다 :)

tributetothemoon replied:
[커밋링크](https://github.com/tributetothemoon/java-lotto/commit/b5471780c4e492f4f08d6e0b0d79378a22d27015)
또한 LottoResult를 LottoTickets가 WinningNumbers를 받아 생성하도록 변경했습니다. LottoResult에서 통계를 계산할 때, LottoTickets에서 lottoTickets()메소드를 통해 List<LottoTicket>을 리턴받아 계산하는 로직의 모양새가 객체에게 묻는 것이 아니라 값을 꺼내와서 계산하는 것 같아 내키지 않았는데, 말씀대로 내부로 이동하니 그런 문제가 사라진 것 같습니다!

dave915 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,43 +1,72 @@
-            return createBudgetMoney();
+    private List<LottoTicket> buyLottosAutomatically(LottoPurchaseManager lottoPurchaseManager) {
+        List<LottoTicket> lottoTickets = new ArrayList<>();
+        while (lottoPurchaseManager.canAfford(LottoTicket.PRICE, 1)) {
```
`1` 매직넘버를 제거해주세요 :)

tributetothemoon replied:
[수정](https://github.com/tributetothemoon/java-lotto/commit/14ba4704ed4c87ffb33e38e9a9e202259231e11a)하였습니다 :) 매직 넘버를 사용하지 않도록 주의하겠습니다. 감사합니다!

dave915 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -8,33 +8,34 @@
+                .map(LottoNumber::from)
                 .filter(lottoNumber -> isNoneDuplicationNumberOrThrow(lottoNumbersList, lottoNumber))
                 .forEach(lottoNumbersList::add);
         return new LottoTicket(lottoNumbersList);
```
이렇게도 쓸 수 있어요 :) 
```
        return lottoNumbers.stream()
                .map(LottoNumber::from)
                .filter(lottoNumber -> isNoneDuplicationNumberOrThrow(lottoNumbersList, lottoNumber))
                .collect(Collectors.collectingAndThen(Collectors.toList(), LottoTicket::new));
```

tributetothemoon replied:
사실 페어 프로그래밍을 할 때 처음에 비슷한 방식으로 콜렉터를 이용해 사용하려고 했었는데요... 스트림문 중간에 중복 검증로직을 넣는 부분에서 막혔던 기억이 납니다.
결국 페어와 고민하다가 미리 어레이리스트를 만들고,
```
List<LottoNumber> lottoNumbersList = new ArrayList<>();
```
중복이 있는지 확인할 리스트와 요소를 받는 static 메소드를 만들어 스트림 문의 중간에 사용하는 방식으로 페어와 합의를 봤는데요.
```
.filter(lottoNumber -> isNoneDuplicationNumberOrThrow(lottoNumbersList, lottoNumber))
```
그래서 제시해주신 방법으로는 중복 확인이 불가능합니다 ㅠㅠ...혹시 더 나은 방법이 있을까요...?

dave915 replied:
stream을 돌기전에 Set등을 이용해서 파라미터로 들어온 `List<Integer>`의 중복체크를 해도 될 것 같아요!

tributetothemoon replied:
[커밋링크](https://github.com/tributetothemoon/java-lotto/commit/e1d5e8f1d9dcaac8f86df925956beb9fe2acceca)
Set를 이용하니 기존의 코드보다 훨씬 간결하게 작성할 수 있게 되었습니다. collectingAndThen의 활용 역시도 배워갑니다. 감사합니다 !

dave915 commented on src/main/java/domain/Rank.java:
```diff
@@ -6,21 +6,21 @@
-    NOTHING(0, new Money("0"));
+    FIRST(6, Money.from(2_000_000_000)),
+    SECOND(5, Money.from(30_000_000)),
+    THIRD(5, Money.from(1_500_000)),
```
```
    SECOND(5, false, new Money("30000000")),
    THIRD(5. true, new Money("1500000"))

    private final long count;
    private final boolean checkBonus;
    private final Money reward;
```

이런식으로 하면 `public static Rank from(long count, boolean bonusBallMatch)` 메소드에서 if부분을 제거 할 수 있지 않을까요?

tributetothemoon replied:
[수정](https://github.com/tributetothemoon/java-lotto/commit/f36be2dbcb9d79744f0619ae1df5d32a9dd5be7e)하였습니다. if문 뿐만 아니라 2등과 3등이 인스턴스 상태의 측면에서 어떻게 다른지까지 나타낼 수 있는 좋은 수단인 것 같습니다. 좋은 피드백 감사합니다!!

dave915 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,49 @@
+        for (LottoTicket lottoTicket : lottoTickets) {
+            Rank rank = winningNumbers.calculateRank(lottoTicket);
+            Integer count = resultStatisticsMap.get(rank);
+            resultStatisticsMap.put(rank, count + 1);
```
`resultStatisticsMap.put(rank, count + 1);` 이 부분을 메소드 분리해서 map에 rank가 없으면 1을 넣어주고 있으면 count + 1 을 해주면 첫번째 for문은 제거 가능할 것 같아요 :)

tributetothemoon replied:
[수정](https://github.com/tributetothemoon/java-lotto/commit/4cb8e54d5b029693ca8a1cfb4ff72dea55622077)하였습니다.
적용하고 보니 for문이 사라지는 것 뿐만 아니라, 당첨되지 않은 등수(rank)는 key와 value 값을 아예 담지 않게 되면서 Map이 담게 되는 데이터 양 역시 줄어들었다는 점이 좋은 것 같습니다. LottoResult는 View단에 보내지는 DTO로서의 역할도 어느 정도 겸하는데, DTO는 가능한 최소 정보만을 담아야한다는 점에서도 좋은 리팩토링을 할 수 있었다고 생각합니다 :) 좋은 피드백 감사합니다.

dave915 reviewed (changes requested):
김김 안녕하세요! 
객체역할에 관한 설명 감사합니다 😁
몇가지 피드백 남겼습니다! 확인 부탁드릴게요.
궁금한 점이 있다면 언제든 DM주세요!

tributetothemoon committed:
refactor: LottoPurchaseManager.canAfford() 파라미터 수정

- 기존에는 구매하고자 하는 재화 1개의 가격과 그 개수를 파라미터로 받았지만, 로또만 대상으로 하므로 굳이 필요 없는 재화 1개의 가격은 파라미터로 받지 않도록 수정하였습니다.

tributetothemoon committed:
style: 매직 넘버 제거

tributetothemoon committed:
refactor: Rank 어트리뷰트에 보너스 번호 일치 여부 추가

- 어트리뷰트에 보너스볼 일치 여부를 추가하여, enum 객체 중 2등과 3등이 상태에 따라 어떻게 구분되는지를 더 확연히 했습니다.
- 또한 Rank.from 메소드 내부의 if 조건절을 삭제할 수 있도록 간략화했습니다.

tributetothemoon committed:
refactor: LottoResult 맵 정보량 축소

- 당첨되지 않은 정보까지는 가지고 있지 않도록 변경하였습니다.

tributetothemoon committed:
feat: LottoTickets 클래스 생성

- LottoTicket들을 담은 일급컬렉션 LottoTickets를 생성하였습니다.

tributetothemoon committed:
refactor: LottoTickets를 이용해 List<LottoTicket>을 Wrapping

tributetothemoon committed:
refactor: LottoTicket의 생성 로직을 Set를 이용하여 리팩토링

tributetothemoon committed:
refactor: LottoResult를 LottoTickets가 WinningNumbers를 받아 생성토록 변경

tributetothemoon commented:
새로운 미션의 페어프로그래밍을 진행하느라 로또 미션의 리팩토링이 늦었습니다😢 늦게 리뷰 요청을 보내드려 죄송합니다😭

dave915 reviewed (approved):
김김! 피드백 반영하신 부분 확인했습니다.
그동안 고생 많으셨습니다 ㅎㅎ 이만 머지할게요!

tributetothemoon commented:
# [JAVA8] 자바 자체 제공 함수형 인터페이스의 활용 - 5
## 내용
- 사용자 입력시 예외가 발생하면 재입력을 요구하는 방식의 코드를 주로 try ~ catch ~ 문을 이용한 재귀식으로 작성하는 패턴을 주로 사용했었다.
- 이때의 문제는 재입력을 요구하고자 하는 함수를 하나 하나 모두 try~ catch 문으로 작성해야 했다는 것이고, 리턴 타입이 다르면 같은 함수로 묶어줄 수도 없었다.
- 이 때, Supplier를 이용해 예외가 발생할 수 있으며, 재입력을 요구하고자 하는 함수를 파라미터로 받아 계속해서 반복되도록 하는 [Repeater](https://github.com/woowacourse/java-lotto/blob/tributetothemoon/src/main/java/util/Repeater.java)라는 클래스를 만들어보았다.
- 비단 이런 유틸 클래스를 만든 것 뿐이 아니라, 자바에서 함수를 파라미터로 받기 위해 자체적으로 제공하는 편리한 제네릭 함수형 인터페이스가 많다는 것을 알게 되었고, 이를 이용해 정확히 어떻게 파라미터로 메소드를 주고 받을 수 있는지를 알게 되었다. 앞으로도 함수형 인터페이스를 적극 활용할 수 있겠다는 생각이 들었다.
## 링크
- [함수형 인터페이스 이해하기](https://codechacha.com/ko/java8-functional-interface/)

-----------

# [ProgramingSkill] 캐싱 - 3
## 내용
- 로또 번호와 같이 범위와 상태가 제한되고, 불변객체라서 캐싱해두면 성능상 유리한 경우, 정적 팩토리 메소드를 이용하면 쉽게 캐싱이 가능하다.
- [해당 예제](https://github.com/woowacourse/java-lotto/blob/tributetothemoon/src/main/java/domain/LottoNumberRepository.java)
